RevenueCatUI/View+PresentPaywall.swift

```diff
@@ -560,14 +681,299 @@ private struct PresentingPaywallModifier: ViewModifier {
             self.restoreFailure?($0)
         }
         .interactiveDismissDisabled(self.purchaseHandler.actionInProgress)
+        .task(id: data.id) {
+            // Reset session purchase result to track purchases in THIS session only.
+            // Using .task(id:) ensures this only runs once per unique session (identified by data.id),
+            // avoiding accidental resets if onAppear is called multiple times.
+            self.purchaseHandler.resetSessionPurchaseResult()
+
+            guard let offering = await self.content.resolveOffering() else { return }
+            self.exitOfferOffering = await ExitOfferHelper.fetchValidExitOffer(for: offering)
+        }
     }
 
+    /// The offering loaded for exit offer presentation
+    @State
+    private var exitOfferOffering: Offering?
+
     private func close() {
         Logger.debug(Strings.dismissing_paywall)
 
         self.data = nil
     }
 
+    private func closeExitOffer() {
+        Logger.debug(Strings.dismissing_paywall)
+
+        self.presentedExitOffer = nil
+        self.exitOfferOffering = nil
+    }
+
+    /// Handles dismissal of the main paywall, checking for exit offers.
+    ///
+    /// We check `purchaseHandler.sessionPurchaseResult` to determine if exit offer should be shown:
+    /// - If a purchase happened in this session, we use `shouldDisplay` with the result's `CustomerInfo`
+    /// - This ensures consistent behavior with how the first paywall decides to show/close
+    private func handleMainPaywallDismiss() {
+        // Don't show exit offer if main paywall is still showing
+        guard self.data == nil else {
+            self.onDismiss?()
+            return
+        }
+
+        // Check shouldDisplay with the purchase result if available
+        if let purchaseResult = self.purchaseHandler.sessionPurchaseResult,
+           !self.shouldDisplay(purchaseResult.customerInfo) {
```

Comment:
do we need to check `shouldDisplay` here if we are dismissing?

Reply:
This is the logic we have in Android on dismissing of the main paywall:
- Check if exit offer
   - If there's an exit offer
      - Check if shouldDisplay indicates to render
         - IF true, show exit offer
         - If false, dismiss   
   - If no intro offer dismiss